packages/eslint-plugin/tests/rules/no-unsafe-type-assertion.test.ts

```diff
@@ -0,0 +1,918 @@
+declare const a: string;
+const b = <boolean>(<string | boolean>a);
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 11,
+          endColumn: 41,
+          data: {
+            type: 'string | boolean',
+            asserted: 'boolean',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+declare const a: string;
+const b = <'bar'>(<'foo'>a);
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 11,
+          endColumn: 28,
+          data: {
+            type: '"foo"',
+            asserted: '"bar"',
+          },
+        },
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 19,
+          endColumn: 27,
+          data: {
+            type: 'string',
+            asserted: '"foo"',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+function f(t: number | string) {
+  return <number | boolean>t;
+}
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 10,
+          endColumn: 29,
+          data: {
+            type: 'string | number',
+            asserted: 'number | boolean',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+function f<T extends number | string>(t: T) {
+  return <number | boolean>t;
+}
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 10,
+          endColumn: 29,
+          data: {
+            type: 'string | number',
+            asserted: 'number | boolean',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+function f<T extends number | string>(t: T) {
+  return <Omit<T, number>>t;
+}
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 10,
+          endColumn: 28,
+          data: {
+            type: 'string | number',
+            asserted: 'Omit<T, number>',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+declare const a: () => string | boolean;
+const b = <() => string | number>a;
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 3,
+          column: 11,
+          endColumn: 35,
+          data: {
+            type: '() => string | boolean',
+            asserted: '() => string | number',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+interface Foo {
+  bar: number;
+  bas: string;
+}
+
+var foo = <Foo>{};
+      `,
+      errors: [
+        {
+          messageId: 'unsafeTypeAssertion',
+          line: 7,
+          column: 11,
+          endColumn: 18,
+          data: {
+            type: '{}',
+            asserted: 'Foo',
+          },
+        },
+      ],
+    },
+    {
+      code: `
+interface Foo {
+  bar: number;
+}
+
+// no additional properties are allowed
```

Comment:
Oh..... Yeah, this may be annoying.

At first glance, at least, I think that this code _should_ be considered valid. I'm guessing that the assignability API surprisingly encodes [excess property checks](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks) into the type system here 🫤.

Note that the following is allowed by the rule.
```ts
interface Foo {
  bar: number;
}

const hmmm = { bar: 1, bazz: 1 }

// no complaint despite additional properties
export const foo = <Foo>hmmm;
```

I wonder if this is a behavior we can control? Maybe we can get the non-object-literal version of the type or something to feed into the assignability check? Not familiar with the details here, but I'd be interested in further investigation and discussion

Reply:
This almost feels like a wholly separate kind of complaint: an _"unnecessary"_ rather than an _"unsafe"_ one. Why would someone want to assert an object typed value to a strictly more narrow object type?

Agreed this rule shouldn't report in this case 😬. It's not really _"unsafe"_ per normal TypeScript type theology IMO. The type system would allow passing `foo` to a location that requires `Foo`, e.g.:

```ts
interface Foo {
  bar: number;
}

const hmmm = { bar: 1, bazz: 1 }

let foo: Foo;
foo = hmmm; // no complaint
```

Reply:
FWIW - Having kicked this around in the back of my head I'm actually starting to think it's probably fine to ban this as long as we have a special case error message. But either way that requires us to be able to detect or infer that that's _why_ it's flagging. If we can achieve that, then we can bikeshed about special-case error messaging, suppressing, or providing an option to control it 🤷

Reply:
Yes! I agree with this. I looked into this a bit more, and I can tell the following works/fails too:

```ts
interface Foo {
  bar: number;
}

// fails
const bar = { bar: 1, bazz: 1 } as Foo;

// works
const foo = { bar: 1, bazz: 1 };
const bar = foo as Foo;
```

The same behavior can be seen on the example on the link @kirkwaiblinger sent ([excess property checks](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks)), as well as assignments with a type annotation (which I assume run `isAssignableTo()` or similar):

```ts
interface Foo {
  bar: number;
}

// fails
const bar: Foo = { bar: 1, bazz: 1 };

// works
const z = { bar: 1, bazz: 1 };
const bazz: Foo = z;
```

I think that, as @kirkwaiblinger explained, TypeScript is making [excess property checks](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks) specifically on object literals. One way around this (as mentioned in the linked article) is assigning the value to a variable.

I'm not sure on this yet, but I think TypeScript may be widening the type as it's assigned to a variable (I'm not sure yet what it means [to widen an object literal](https://github.com/microsoft/TypeScript/blob/e99e6e2e87a6d7f152eb7cc491ce11da1926662b/src/compiler/checker.ts#L25212-L25228)). I can verify that `isTypeAssignableTo()` with the widened type (via `getWidenedType()`, specifically if the type is an object literal), works as you'd expect, and doesn't fails on excess properties.